specification/cognitiveservices/data-plane/TranslatorText/Batch/preview/v1.0/TranslatorBatch.json

```diff
@@ -0,0 +1,1469 @@
+              "application/json": "{\r\n  \"error\": {\r\n    \"code\": \"InvalidRequest\",\r\n    \"message\": \"Some argument is incorrect\",\r\n    \"target\": \"operation\",\r\n    \"innerError\": {\r\n      \"code\": 400120,\r\n      \"error\": \"SASTokenInvalid\",\r\n      \"message\": \"SAS token for storage is invalid\"\r\n    }\r\n  }\r\n}"
+            }
+          },
+          "401": {
+            "description": "Unauthorized.  Please check your credentials",
+            "schema": {
+              "$ref": "#/definitions/ErrorResponseV2"
+            },
+            "examples": {
+              "application/json": "{\r\n  \"error\": {\r\n    \"code\": \"Unauthorized\",\r\n    \"message\": \"User is not authorized\",\r\n    \"target\": \"operation\",\r\n    \"innerError\": {\r\n      \"code\": 401000,\r\n      \"error\": \"Unauthorized\",\r\n      \"message\": \"Operation is not authorized\"\r\n    }\r\n  }\r\n}"
+            }
+          },
+          "429": {
+            "description": "Too many requests",
+            "schema": {
+              "$ref": "#/definitions/ErrorResponseV2"
+            },
+            "examples": {
+              "application/json": "{\r\n  \"error\": {\r\n    \"code\": \"RequestRateTooHigh\",\r\n    \"message\": \"User\\u0027s request rate is too high\",\r\n    \"target\": \"operation\",\r\n    \"innerError\": {\r\n      \"code\": 429000,\r\n      \"error\": \"RateTooHigh\",\r\n      \"message\": \"Request rate is too high\"\r\n    }\r\n  }\r\n}"
+            }
+          },
+          "500": {
+            "description": "Internal Server Error",
+            "schema": {
+              "$ref": "#/definitions/ErrorResponseV2"
+            },
+            "examples": {
+              "application/json": "{\r\n  \"error\": {\r\n    \"code\": \"InternalServerError\",\r\n    \"message\": \"Internal Server Error\",\r\n    \"target\": \"operation\",\r\n    \"innerError\": {\r\n      \"code\": 500000,\r\n      \"error\": \"Internal Server Error\",\r\n      \"message\": \"Unexpected internal server error has occurred\"\r\n    }\r\n  }\r\n}"
+            }
+          },
+          "503": {
+            "description": "Server temporary unavailable",
+            "schema": {
+              "$ref": "#/definitions/ErrorResponseV2"
+            },
+            "examples": {
+              "application/json": "{\r\n  \"error\": {\r\n    \"code\": \"ServiceUnavailable\",\r\n    \"message\": \"Service is temporary unavailable\",\r\n    \"target\": \"operation\",\r\n    \"innerError\": {\r\n      \"code\": 503000,\r\n      \"error\": \"ServiceTemporaryUnavailable\",\r\n      \"message\": \"Service is currently unavailable.  Please try again later\"\r\n    }\r\n  }\r\n}"
+            }
+          }
+        },
+        "x-ms-examples": {
+          "Get status for operation list": {
+            "$ref": "./examples/operations.json"
+          }
+        }
+      }
+    },
+    "/v1.0-preview.1/batches/upload": {
```

Comment:
I feel that the name upload is problematic here, because it doesn't indicate the following salient aspects of this endpoint:

1. A single document is being uploaded
2. The upload takes a binary format rather than the JSON description used for /batches
3. The storage for source and target are hosted by the service.

In CustomVision, they provide separate endpoints for images uploaded as binary (/classify/iterations/{name}/images), and provided as a URL (/classify/iterations/{name}/url).  In Computer Vision and Form Recognizer, they use the same endpoint, and differentiate URL vs. binary via the Content-Type.  However, for all of these cases, there is a single file to process.

In this case, since the operation still happens as a batch -- of one -- I'd consider renaming it to single or singleHosted.  Alternatively, you could look at a way to make this a true batch of uploaded binary documents to parallel /batches, and call it /batches/hosted.

Reply:
I agree.  Let's see what the user feedback says.  We will go with /batches/single for now.
Or how about /batches/singleBinary?

Reply:
or /batches/single/binary ?

Reply:
It is relatively straight-forward to overload based on different kinds of input. The main question to me is if we can unify the outputs.

Reply:
It is not straightforward because the single / upload case is microsoft hosted storage vs. the other case is customer hosted storage.
So for the single / upload case, we need to return the url for the translated doc (which will include a SAS token and another field that indicates an expiration date).
For the batches case, we don't need any of that since customer owns the destination storage.

Reply:
The outputs here are the results of LROs -- specifically the `OperationStatusResponse` which gives the status of the full operation as a group.  You have to query the /batches/{id}/documents/{documentId} endpoint to get the document path, and the SAS token to access it from service-hosted storage isn't available here, so callers who aren't the document owners can't access it.  If there were a way to provide the SAS token from either the `OperationStatusResponse` or the `DocumentStatusResponse`, then you could unify these endpoints.

Reply:
I'd prefer making the endpoint support a batch by accepting multipart form data as Johan suggested.  It could then be renamed /batches/hosted and be a proper batch operation, which would still work for a batch of one.

Reply:
Only the azure subscriber will have access to the batch / document status.  But there are scenarios where the same subscription can be used by multiple users and those users probably shouldn't have access to other users actual translated document.

Reply:
some more challenges if we do multipart form data approach:
1. if we do the formdata approach, how to you map the multiple document -> actual document ids.  example, if i have 3 documents binary in the POST, i'll get back a batchid, and then i query for the document status but then i can't figure out which is which.

2. how can we specify the document type (like docx vs. text vs. xlif), right now we use the contentype.  Would we require an extra field in the formdata to specify the format type for each document?
​
3. right now for the batches/single we use the query string parameters to specify source and target lanugage, should we keep that in form data implementation as well or should that move inside to the body?

Reply:
ok looks like we can use multipart/formdata that includes the filename
https://docs.microsoft.com/en-us/aspnet/core/mvc/models/file-uploads?view=aspnetcore-3.1 and contenttype